### PR TITLE
Qwen2.5 Coder 32B bringup to TTTv2

### LINKED ISSUE #44691: Qwen2.5-Coder-32B-Instruct bringup to TTTv2

### PR DESCRIPTION
**Pull request by adrian-pascual-bernal**
Closes #44691
# [TTTv2] Qwen2.5-Coder-32B bring-up (T3K)

## TLDR

> **UPDATE 2026-07-03 (healthy build, submodules synced).** The 2026-06-17 numbers below were
> measured on a stale-build worktree (out-of-date submodules) — they are ~30-45% low (same
> stale-build effect seen on Qwen2.5-72B). Re-measured **batch-1**
> on a synced/healthy build (`umd 73d9ccd`), reset-then-measure:
> - **perf · batch-1 · on-device (`on_device_topk`) = 23.3 t/s/u** (43 ms/tok) vs real **TTTv1 = 23.97** (same build) → **parity**.
> - **accuracy · batch-1 · on-device = 20.2 t/s/u** (49.6 ms/tok) — **no longer hangs**.
> - token-accuracy: perf **95.3 / 99.6**, accuracy **98.6 / 100.0**.
>
> **Both `batch-1 · on-device` hangs are now FIXED.** The *accuracy* hang was cleared earlier by the bf8-KV recipe
> (`model.py` `kv_cache_dtype`); the *performance* `on_device_topk` hang — a distinct
> sampling-buffer-allocated-during-trace bug — was fixed by the shared
> `TracedLLMExecutor._prealloc_sampling_buffers` change (= PR #48900, kept byte-identical on this branch). The old
> "gap vs 19.7/22.4" was a host-path + stale-PERF.md artifact; on-device it is at TTTv1 parity.
>
> **Still stale / re-measure pending on the healthy build:** all `batch-32` cells, all `on-host` cells, and TTTv1
> `accuracy · batch-1`. NOTE: this T3K box is bimodal — measure perf only after a fresh `tt_device_reset` (a no-reset
> run gave a spurious 10.3 t/s/u on the same on-device code path).

TTTv2 port of `Qwen/Qwen2.5-Coder-32B-Instruct` (T3K only). Prefill trace seq-lens `[128, 1024]`. (The original
2026-06-17 TLDR — "one cell hangs, others at parity" — is superseded by the banner above.)

## What changed

The shared TTTv2 executor, the `Sampling1D` sampling infra, and the force-argmax fix are **already merged to main**. After rebasing onto main, this PR contains only the Qwen2.5-Coder-32B model-specific commits:

- Add Qwen2.5-Coder-32B-Instruct TTTv2 port (T3K only; the demo `pytest.skip`s any non-T3K mesh).
- M4 accuracy/perf demo + the two `PrecisionConfig` recipes (`QWEN25_CODER_32B_ACCURACY`, `QWEN25_CODER_32B_PERFORMANCE`).
- Mirror TTTv1's non-precision config: prefill chunk cutoff (`mlp_prefill_len_cutoff=1024`), perf-mode HIFI2 attention compute kernel, and `max_prefill_chunk_size=4096`.

## Sampling note

- **on-host:** argmax (shipped default; `sampling_params=None`).
- **on-device:** top-k path, but in practice (AFAIK) effectively argmax, so it is sampling-equivalent to the host argmax path.

`Sampling1D` is constructed in the model with `allow_force_argmax=False` (the shared-base fix gates the force-argmax collapse on this flag; this model does not need it flipped). `FABRIC_1D` is enabled on the >1-device mesh and a `SAMPLING_MODE` env toggle selects the path.

## Prefill note

Prefill is **not batched** — TTTv2 runs prefill sequentially per user. For an apples-to-apples TTFT comparison, the TTTv1 TTFT figures here use TTTv1's sequential prefill (`DISABLE_BATCHED_PREFILL=1`), which is exactly the per-user-sequential prefill TTTv2 does. No TTTv1 batched-prefill TTFT is reported. Prefill trace capture is enabled for seq-lens `[128, 1024]` on T3K (TTTv1 device default; confirmed on hardware with no TT_FATAL).

## Accuracy

Book top1/top5 (teacher-forcing), graded against a freshly generated book reference (HF reference intrinsic ceiling 89.5 / 97.27 top1/top5 over the full next-token positions; the demo half-splits at `prompt_len=512`).

| SKU | Profile | TTTv2 top1/top5 | Reference top1/top5 | Verdict |
|---|---|---|---|---|
| T3K | performance | 95.3 / 99.6 | 89.5 / 97.27 | PASSED |
| T3K | accuracy | 98.6 / 100.0 | 89.5 / 97.27 | PASSED |

_(top1/top5 re-measured 2026-07-03, healthy build; within noise of the 2026-06-17 95.1/99.6 · 98.2/100.0.)_

### Performance profile

| SKU | batch | TTTv1 TTFT (ms) | TTTv1 t/s/u | TTTv2 on-host TTFT (ms) | TTTv2 on-host t/s/u | TTTv2 on-device TTFT (ms) | TTTv2 on-device t/s/u |
|---|---|---|---|---|---|---|---|
| T3K | batch-1 | 95.7 | 23.97 | 95.9 | 9.3 | 102.4 | 23.3 |
| T3K | batch-32 | 92.4 | 11.0 | 92.8 | 8.5 | 92.6 | 15.9 |

The on-device column is the SKU-matched comparison vs TTTv1 on T3K (≥8-dev). batch-32 TTTv1 TTFT is the sequential value (`DISABLE_BATCHED_PREFILL=1`); batch-1 TTTv1 TTFT is used directly.

### Accuracy profile

| SKU | batch | TTTv1 TTFT (ms) | TTTv1 t/s/u | TTTv2 on-host TTFT (ms) | TTTv2 on-host t/s/u | TTTv2 on-device TTFT (ms) | TTTv2 on-device t/s/u |
|---|---|---|---|---|---|---|---|
| T3K | batch-1 | 111.7 | 12.7 | 117.7 | 8.2 | 118.1 | 20.2 |
| T3K | batch-32 | 106.3 | 9.1 | 106.4 | 7.9 | 106.2 | 12.4 |

The on-device column is the SKU-matched comparison vs TTTv1 on T3K (≥8-dev). batch-32 TTTv1 TTFT is the sequential value (`DISABLE_BATCHED_PREFILL=1`); batch-1 TTTv1 TTFT is used directly.

### Summary vs TTTv1 (fastest TTTv2 mode per cell)

The single **TTTv2** column — and both **Δ%** columns — use the *fastest* TTTv2 sampling mode for each cell (better of on-host / on-device above, by decode t/s/u; on T3K that is on-device). Δ% TTFT = (TTTv2 − TTTv1)/TTTv1 × 100 (negative = TTTv2 faster); Δ% t/s/u = same (positive = TTTv2 faster). batch-32 TTTv1 TTFT = sequential (`DISABLE_BATCHED_PREFILL=1`) value; batch-1 direct.

#### Performance profile

| SKU | batch | TTTv1 TTFT (ms) | TTTv1 t/s/u | TTTv2 TTFT (ms) | TTTv2 t/s/u | Δ% TTFT | Δ% t/s/u |
|---|---|---|---|---|---|---|---|
| T3K | batch-1 | 95.7 | 23.97 | 102.4 | 23.3 | +7.0 | -2.8 |
| T3K | batch-32 | 92.4 | 11.0 | 92.6 | 15.9 | +0.2 | +44.5 |

#### Accuracy profile

| SKU | batch | TTTv1 TTFT (ms) | TTTv1 t/s/u | TTTv2 TTFT (ms) | TTTv2 t/s/u | Δ% TTFT | Δ% t/s/u |
|---|---|---|---|---|---|---|---|
| T3K | batch-1 | 111.7 | 12.7 | 118.1 | 20.2 | —¹ | —¹ |
| T3K | batch-32 | 106.3 | 9.1 | 106.2 | 12.4 | -0.1 | +36.3 |

**[2026-07-03]** The `accuracy · batch-1 · on-device` hang is **FIXED** (see Caveats), so the TTTv2 column now
uses the on-device path (**20.2 t/s/u**, was the host-fallback 8.2). ¹ Δ% is left blank because the TTTv1
`accuracy · batch-1` baseline (12.7 / 111.7) is a 2026-06-17 **stale-build** number, not re-measured on the healthy
build — re-measure TTTv1 accuracy·b1 before computing Δ. (For reference, all batch-1 **performance**-profile cells
and the perf summary Δ above ARE on the healthy build.)

## Caveats

- **~~`accuracy · batch-1 · on-device` hangs deterministically (T3K).~~ RESOLVED (2026-07-03).** Both
  `batch-1 · on-device` hangs are fixed. (1) *accuracy · batch-1*: cleared by the bf8-KV accuracy recipe
  (`model.py` `kv_cache_dtype`) — now runs at **20.2 t/s/u**. (2) *performance · batch-1 ·
  `on_device_topk`*: a **distinct** decode-replay hang found this session (froze right after `Captured decode
  trace`, 0 decode steps) — root cause was on-device sampling buffers (Sampling1D device buffers + k/p/temp)
  allocated *during trace-active state* and clobbered on replay; fixed by `TracedLLMExecutor._prealloc_sampling_buffers`
  called before the prefill trace (= PR #48900, kept byte-identical on this branch). Now runs at **23.3 t/s/u**.
  (The original bf8-KV "fix" for the accuracy hang was in hindsight a layout band-aid for this same class of bug;
  retained on its own merits — loss-free + lighter decode traffic.)

## Repro

```bash
# T3K only. All ttnn runs go through the device queue with
# PYTHON_ENV_DIR=<worktree>/python_env (never raw Bash).
export MESH_DEVICE=T3K HF_MODEL=Qwen/Qwen2.5-Coder-32B-Instruct

# TTTv2 (SAMPLING_MODE = host | on_device | on_device_topk); profile = performance | accuracy
SAMPLING_MODE=on_device_topk pytest models/common/tests/demos/qwen25_coder_32b/demo.py -v \
  -k 'performance and batch-1'    # also: 'performance and batch-32', 'accuracy and batch-32'
# accuracy · batch-1 · on-device HANGS — use SAMPLING_MODE=host for that single cell.

# On-device prefill-trace confirmation (look for "Captured prefill trace for seq_len=128", no TT_FATAL):
pytest models/common/tests/demos/qwen25_coder_32b/demo.py -v -k 'performance and batch-32'

# Token accuracy (book; top1/top5 logged before the threshold assert):
pytest models/common/tests/demos/qwen25_coder_32b/demo.py -v -k 'token-accuracy'

# TTTv1 baseline (auto on-device sampling, batched prefill):
pytest models/tt_transformers/demo/simple_text_demo.py -v -k 'performance and batch-1'
# TTTv1 batch-32 apples-to-apples (sequential per-user prefill):
DISABLE_BATCHED_PREFILL=1 pytest models/tt_transformers/demo/simple_text_demo.py -v \
  -k 'performance and batch-32 and not probs'
```

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:apascual/44691-qwen25-32b-coder-bringup-tttv2)

### CODE REVIEW

**Comment by adrian-pascual-bernal** [comment]
As can be observed, there is a substantial decode-throughput gap between this TTTv2 port and TTTv1 (TTTv2 lands at ~47% of the PERF.md target on T3K). A full report summarising the profiling work — both device-side (Tracy op traces) and Python-level instrumentation of the executor decode loop — is attached.

The gap is not a device-compute or precision-recipe regression. It is host-side overhead in the per-decode-step Python executor loop, dominated by the path that materialises the full [B, V] logits tensor on the host between decode steps (per-shard ttnn.to_torch + torch.cat stitch followed by host argmax). The cost of this stitch scales with vocab_size × num_TP_shards, and Qwen2.5-Coder-32B on T3K hits both factors hard (vocab = 152,064, 8-way TP).

To confirm this attribution, the same profiling methodology was applied to the Mistral-7B-Instruct-v0.3 TTTv2 port on N300 — which exercises the identical executor code path. Mistral sits at ~97% of its PERF.md target, with the same _process_output_decode stitch costing ~0.8 ms / step (vs ~23 ms / step on Qwen-Coder-32B). This rules out a TTTv2 executor regression and pins the gap to a vocab × shard-count-driven host cost.

The follow-up workstream is wiring on-device sampling (Sampling1D) so that only a [B, 1] token tensor is read back per step instead of full-vocab logits — this is the same path TTTv1 uses for its published numbers.

[PERF_GAP_REPORT.md](https://github.com/user-attachments/files/28357160/PERF_GAP_REPORT.md)

File: models/common/models/qwen25_coder_32b/weights.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Host-side weight helpers for Qwen2.5-Coder-32B-Instruct TTTv2.
+
+Construction lives in
+:func:`models.common.models.qwen25_coder_32b.model.Qwen25Coder32B.from_pretrained`;
+low-level permutes and HF tensor layout live in ``weight_utils``.
+"""
+
+from models.common.models.qwen25_coder_32b import weight_utils
+
+__all__ = ["weight_utils"]
```
Comment (gwangTT) [inline models/common/models/qwen25_coder_32b/weights.py]: pointless file?

**Review by gwangTT — APPROVED** [review]

**Review by mtairum — APPROVED** [review]

**Comment by yieldthought** [comment]
Codex review:

I found a few issues that look blocking for landing this as a promoted TTTv2 demo:

1. [models/common/models/qwen25_coder_32b/model.py:94](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/qwen25_coder_32b/model.py#L94) enables prefill tracing whenever the new prefill length is 128 or 1024, even when `num_cached_tokens > 0`. In the shared traced executor, that path calls `_easy_trace_prefill(...)` without carrying the cached start position or chunk start into the trace body ([models/common/models/executor.py:1248](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/executor.py#L1248) and [models/common/models/executor.py:1266](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/executor.py#L1266)), while the eager prefix-cache path only gets the correct context by passing `chunk_start_idx` into `model.prefill_forward` ([models/common/models/executor.py:748](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/executor.py#L748)). As a result, a prefix-cached prefill with 128 or 1024 new tokens can take the trace path and compute attention over just the new chunk instead of the existing KV prefix plus the new tokens. Please keep this trace policy to cold prefill (`num_cached_tokens == 0`) until traced prefix-cache prefill passes the chunk start through.

2. [models/common/tests/demos/qwen25_coder_32b/demo.py:60](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/tests/demos/qwen25_coder_32b/demo.py#L60) hard-codes the old PERF.md throughput targets for every performance demo, and the parametrized test matrix at [models/common/tests/demos/qwen25_coder_32b/demo.py:375](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/tests/demos/qwen25_coder_32b/demo.py#L375) asserts those targets at [models/common/tests/demos/qwen25_coder_32b/demo.py:568](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/tests/demos/qwen25_coder_32b/demo.py#L568). The PR description reports the best TTTv2 decode throughput as 16.2/15.9 tok/s/u for the performance profile and 8.2/12.4 tok/s/u for the accuracy profile, with `accuracy / batch-1 / on-device` still hanging; all of those are below the 22.4/19.7 PERF.md targets copied here. So the documented demo commands either fail the assertion or hang for the known on-device accuracy/batch-1 case. Please update the expected metrics to the accepted TTTv2 cell-specific targets or parity criterion, and skip/xfail the known-hanging sampling mode rather than landing a promoted demo matrix that cannot pass.

3. [models/common/models/qwen25_coder_32b/generator.py:87](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/qwen25_coder_32b/generator.py#L87) silently turns `warmup_model_prefill` into a no-op whenever the executor wrapper lacks that method. The only traced wrapper constructed by this generator does not expose `warmup_model_prefill`, even though the underlying `TracedLLMExecutor` implements it at [models/common/models/executor.py:982](https://github.com/tenstorrent/tt-metal/blob/574bb26bc8f60f318bfa687499732fb0d8576f50/models/common/models/executor.py#L982). A caller that asks the generator to warm prefill traces therefore gets no warmup, so the first request still captures traces at runtime. Please expose `warmup_model_prefill` on `TracedQwen25Coder32BExecutor` and delegate unconditionally from the generator, or raise for unsupported eager mode instead of silently returning `None`.
